### PR TITLE
eth, les, cmd: limit LES connections, add ethpeers flag

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR adds peer count limitation in les.ProtocolManager. It also adds a new "ethpeers" flag because the number of ETH peers has to be limited too in order to leave some p2p slots open for LES and the previously existing "magic" calculation (MaxPeers-LightPeers unless it's less than half of MaxPeers) was an ugly (and also undocumented) solution with potentially unexpected outcomes for the user. Now that we have multiple protocols, we need per-protocol peer count limitation too.
Global p2p.Config.MaxPeers is set to the value of "maxpeers" flag only if explicitly specified, otherwise it is set to "ethpeers"+"lightpeers" if either of those is specified. This ensures that global peer limit will not cause such a situation by accident where one of the protocols cannot connect because the other has too many connections open.

### CODE REVIEW

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -179,7 +184,7 @@ var (
 	LightPeersFlag = cli.IntFlag{
 		Name:  "lightpeers",
 		Usage: "Maximum number of LES client peers",
-		Value: 20,
+		Value: 100,
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply cmd/utils/flags.go]: I have found that 20 was a very conservative limit (it was just a wild guess). Clients just sync new headers most of the time, I think we can allow a lot more to be connected.

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -800,7 +805,10 @@ func SetP2PConfig(ctx *cli.Context, cfg *p2p.Config) {
 
 	if ctx.GlobalIsSet(MaxPeersFlag.Name) {
 		cfg.MaxPeers = ctx.GlobalInt(MaxPeersFlag.Name)
+	} else if ctx.GlobalIsSet(EthPeersFlag.Name) || ctx.GlobalIsSet(LightPeersFlag.Name) {
+		cfg.MaxPeers = ctx.GlobalInt(EthPeersFlag.Name) + ctx.GlobalInt(LightPeersFlag.Name)
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply cmd/utils/flags.go]: We have a protocol-level limit for each protocol and a global limit at p2p level. In the current logic, the total limit is the sum of the per-protocol limits by default. Still, you can set a total limit lower than this sum if you really want to. I am not sure this is the best solution but at least this ensures that the total limit won't screw up the nodes of unaware users by not letting one of the protocols get peers at all.

File: cmd/utils/flags.go [hunk 3]
```diff
@@ -171,6 +171,11 @@ var (
 		Value: &defaultSyncMode,
 	}
 
+	EthPeersFlag = cli.IntFlag{
+		Name:  "ethpeers",
+		Usage: "Maximum number of ETH peers",
+		Value: 25,
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply cmd/utils/flags.go]: right, done

**Comment by holiman** [comment]
I added the following:

```diff
#git diff
diff --git a/cmd/utils/flags.go b/cmd/utils/flags.go
index 2dc077a..f63e062 100644
--- a/cmd/utils/flags.go
+++ b/cmd/utils/flags.go
@@ -842,6 +842,8 @@ func SetP2PConfig(ctx *cli.Context, cfg *p2p.Config) {
                cfg.NoDiscovery = true
                cfg.DiscoveryV5 = false
        }
+       log.Info("P2P config","MaxPeers", cfg.MaxPeers)
+
 }
 
 // SetNodeConfig applies node-related command line flags to the config.
@@ -1082,6 +1084,8 @@ func SetEthConfig(ctx *cli.Context, stack *node.Node, cfg *eth.Config) {
        if gen := ctx.GlobalInt(TrieCacheGenFlag.Name); gen > 0 {
                state.MaxTrieCacheGen = uint16(gen)
        }
+       log.Info("ETH config","EthPeers", cfg.EthPeers, "LightPeers", cfg.LightPeers)
+
 }
 
 // SetDashboardConfig applies dashboard related command line flags to the config.
```

I don't make sense of this... If I specify `maxpeers`, shouldn't that place a bound on the sub-protocol peers? 

```
#build/bin/geth --maxpeers 5
INFO [01-23|09:06:02] P2P config                               MaxPeers=5
INFO [01-23|09:06:02] ETH config                               EthPeers=25 LightPeers=100

#build/bin/geth --maxpeers 5 --lightpeers 10 --ethpeers 5
INFO [01-23|09:06:50] P2P config                               MaxPeers=5
INFO [01-23|09:06:50] ETH config                               EthPeers=5 LightPeers=10

#build/bin/geth --maxpeers 0 
INFO [01-23|09:07:27] P2P config                               MaxPeers=0
INFO [01-23|09:07:27] ETH config                               EthPeers=25 LightPeers=100
```

**Comment by zsfelfoldi** [comment]
@holiman maybe you are right and my approach was very counterintuitive. What I did was that I set MaxPeers to EthPeers+LightPeers by default and overwrite default MaxPeers if --maxpeers is not specified but either --ethpeers or --lespeers is. If I understand correctly, your approach is to decrease EthPeers and/or LesPeers (which one?) if a MaxPeers smaller than the sum of the two is specified. I am not sure which one is better, probably both can be confusing. I am also not sure whether we still really need the total limit if we have per-protocol limits. We had a plan for a better peer connection API that could handle this more elegantly but I need some acceptable solution now.

**Comment by holiman** [comment]
So, I'd say the following.. 

* If all three are specified, and conflicting, all bets are off. 
* If two flags are specified, the third adopts according to those two. 
* If only maxpeers is specified, the two others are either capped/increased according to some ratio, or possibly ethpeers get's precedence slots up to default, then fills up with lightpeers. 
* More specialized flags (ethpeers / lightpeers) takes precedence over general flags (maxpeers).

**Comment by holiman** [comment]
Question, though; do we even _need_ `maxpeers` ? I only ever use it when I want to try out something without a real blockchain, I do `--maxpeers=0 --nodiscover`. For that usecase, it'd be better with a `--hermit`-flag. 

For any real usecase, it would be better if people actually specified number of `ethpeers` and `lightpeers` instead of using `maxpeers`.

**Comment by zsfelfoldi** [comment]
I'd gladly get rid of maxpeers, the only issue with that is that is has been there forever and we usually don't remove flags overnight :) I like your approach of "specify any 2 of 3". Maybe I should also check which protocol(s) are we actually using.

**Comment by zsfelfoldi** [comment]
Here is my latest best idea:
```
	ethPeers := ctx.GlobalInt(EthPeersFlag.Name)
	ethPeersFixed := ctx.GlobalIsSet(EthPeersFlag.Name)
	lightPeers := ctx.GlobalInt(LightPeersFlag.Name)
	lightPeersFixed := ctx.GlobalIsSet(LightPeersFlag.Name)
	if ctx.GlobalBool(LightModeFlag.Name) {
		// ETH is not used
		ethPeers = 0
		ethPeersFixed = true
	}
	if !ctx.GlobalBool(LightModeFlag.Name) && ctx.GlobalInt(LightServFlag.Name) == 0 {
		// LES is not used
		lightPeers = 0
		lightPeersFixed = true
	}

	if ctx.GlobalIsSet(MaxPeersFlag.Name) {
		maxPeersDiff := ctx.GlobalInt(MaxPeersFlag.Name) - ethPeers - lightPeers
		if maxPeersDiff > 0 {
			if !ethPeersFixed {
				ethPeers += maxPeersDiff
			} else if !lightPeersFixed {
				lightPeers += maxPeersDiff
			}
		} else if maxPeersDiff < 0 {
			if !lightPeersFixed {
				if lightPeers >= -maxPeersDiff {
					lightPeers += maxPeersDiff
					maxPeersDiff = 0
				} else {
					maxPeersDiff += lightPeers
					lightPeers = 0
				}
			}
			if !ethPeersFixed {
				if ethPeers >= -maxPeersDiff {
					ethPeers += maxPeersDiff
					maxPeersDiff = 0
				} else {
					maxPeersDiff += ethPeers
					ethPeers = 0
				}
			}
		}
	}
	cfg.MaxPeers = ethPeers + lightPeers
```

File: cmd/utils/flags.go [hunk 4]
```diff
@@ -992,13 +998,67 @@ func SetEthConfig(ctx *cli.Context, stack *node.Node, cfg *eth.Config) {
...
+	ethPeers := cfg.EthPeers
+	if ctx.GlobalIsSet(EthPeersFlag.Name) {
+		ethPeers = ctx.GlobalInt(EthPeersFlag.Name)
+	}
+	ethPeersFixed := ctx.GlobalIsSet(EthPeersFlag.Name)
+	lightPeers := cfg.LightPeers
+	if ctx.GlobalIsSet(LightPeersFlag.Name) {
+		lightPeers = ctx.GlobalInt(LightPeersFlag.Name)
+	}
+	lightPeersFixed := ctx.GlobalIsSet(LightPeersFlag.Name)
```
Comment (holiman) [inline cmd/utils/flags.go]: Nitpick, I'd write it as 
```golang
	ethPeers := cfg.EthPeers
	lightPeers := cfg.LightPeers

	ethPeersFixed := ctx.GlobalIsSet(EthPeersFlag.Name)
	if ethPeersFixed {
		ethPeers = ctx.GlobalInt(EthPeersFlag.Name)
	}
	lightPeersFixed := ctx.GlobalIsSet(LightPeersFlag.Name)
	if lightPeersFixed {
		lightPeers = ctx.GlobalInt(LightPeersFlag.Name)
	}
```

**Comment by holiman** [comment]
With the new logic: 
```
#build/bin/geth --maxpeers 5
INFO [01-25|14:43:38] ETH config                               EthPeers=5 LightPeers=0
INFO [01-25|14:43:38] P2P config                               MaxPeers=5

#build/bin/geth --maxpeers 5 --lightserv 1
INFO [01-25|14:46:41] ETH config                               EthPeers=5 LightPeers=0
INFO [01-25|14:46:41] P2P config                               MaxPeers=5

#build/bin/geth --maxpeers 25 --lightserv 1
INFO [01-25|14:47:00] ETH config                               EthPeers=25 LightPeers=0
INFO [01-25|14:47:00] P2P config                               MaxPeers=25

#build/bin/geth --maxpeers 15 --ethpeers 10 
INFO [01-25|14:43:45] ETH config                               EthPeers=10 LightPeers=0
INFO [01-25|14:43:45] P2P config                               MaxPeers=10

#build/bin/geth --maxpeers 15 --ethpeers 10 --lightserv 1
INFO [01-25|14:46:13] ETH config                               EthPeers=10 LightPeers=5
INFO [01-25|14:46:13] P2P config                               MaxPeers=15

#build/bin/geth --maxpeers 0 
INFO [01-25|14:43:50] ETH config                               EthPeers=0 LightPeers=0
INFO [01-25|14:43:50] P2P config                               MaxPeers=0

#build/bin/geth --ethpeers 5 --lightserv 1 
INFO [01-25|14:47:37] ETH config                               EthPeers=5 LightPeers=100
INFO [01-25|14:47:37] P2P config                               MaxPeers=105

```

It seems that it's pretty hard to make it accept any lightpeers, may be that users find it difficult to configure it correctly? 

I'd suggest that you also add the following to make it less opaque:
```diff

diff --git a/cmd/utils/flags.go b/cmd/utils/flags.go
index f46ea44..6616d0c 100644
--- a/cmd/utils/flags.go
+++ b/cmd/utils/flags.go
@@ -1121,6 +1121,8 @@ func SetEthConfig(ctx *cli.Context, stack *node.Node, cfg *eth.Config, p2pcfg *p
        if gen := ctx.GlobalInt(TrieCacheGenFlag.Name); gen > 0 {
                state.MaxTrieCacheGen = uint16(gen)
        }
+       log.Info("ETH config","EthPeers", cfg.EthPeers, "LightPeers", cfg.LightPeers)
+       log.Info("P2P config","MaxPeers", p2pcfg.MaxPeers)
 }
```

**Comment by lmars** [comment]
Is it worth adding a test covering the desired behaviour for each flag combination? I'd really like it to be intuitive, the flag combinations are currently a bit of a minefield.

**Comment by zsfelfoldi** [comment]
@lmars I'm not sure how you can unit test intuitiveness :) I think in case of ETH nodes and light clients (where only one kind of peers is allowed) it is pretty straightforward. In case of light servers (where both ETH and LES peers are needed) it is a bit more complicated. Maybe we could print a log info message with the current peer limits in this case.

**Comment by holiman** [comment]
Well, regardless if it's intuitive or not, please add the log statement so users can tell if they've successfully managed to set it up as they wanted to.

**Comment by zsfelfoldi** [comment]
@holiman done

**Comment by lmars** [comment]
@zsfelfoldi 

> I'm not sure how you can unit test intuitiveness

I guess I rolled two points into one, sorry.

My first point is that I'd like to see some tests which serve both to make sure the flags do have the desired behaviour and also ensure we don't regress in the future.

My second point was that I would like the flags to be intuitive, and I actually thought @holiman was highlighting some non-intuitive combinations which seem like undesired behaviour.

For example, if I set `--maxpeers 25 --lightserv 1`, is it not reasonable to assume I'll accept some light peers?

These are just my desires though, feel free to just ship the change if it fixes the [original issue](https://github.com/ethereum/go-ethereum/issues/15689), I just want to get back to serving light peers :wink:.

**Comment by karalabe** [comment]
So, we've had a review call on this and the decision was to not add the `--ethpeers` flag since it's a very messy interplay between `--ethpeers`, `--lightpeers` and `--maxpeers`.

The ugly issue deep in the `p2p` code is that peers are globally limited, and they are also 50% dialed ones and 50% inbound ones. Even with very funky peer counters, `eth` can be somewhat starved by `les` simply by filling up the inbound peer pool and leaving `eth` to dial for `eth` servers itself.

The only meaningful long term solution would be to completely split off peer limits based on subprotocols, not globally in `p2p`. That would entail completely dropping `--maxpeers` and reworking the peer counter logic. If we're going to rework how users specify peer limits, it doesn't make sense to add a bunch of new flags now that get changed in a followup release anyway.

The immediate issue this PR aims to solve is to limit the light connections to `--lightpeers`. That can be done without introducing the `--ethpeers` flag. If the use doesn't supply `--maxpeers`, but does supply `--lightpeers=X`, we can safely assume `maxpeers == default-maxpeers + X`. If the user explicitly supplied `--maxpeers`, we can honor it. If `--maxpeers <= --lightpeers`, we just refuse to start.

The above logic preserves the current behavior of the flags, the current expected behavior of the flags, fixes the immediate starvation issue and doesn't add any new API surface. We can then think of the more extensive fix to track peers on each protocol. Closing this PR in favor of the simplified approach mentioned above.

**Pull request closed without merging**